### PR TITLE
Running dataclass transform in a later pass to fix crashes

### LINKED ISSUE #12685: Crash
Running MyPy 0.950 on [the latest efax 1.6.3](https://github.com/NeilGirdhar/efax) gives
```shell
❯ mypy efax --show-traceback
efax/_src/distributions/dirichlet.py:29: error: INTERNAL ERROR -- Please try using mypy master on Github:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 0.950
Traceback (most recent call last):
  File "mypy/semanal.py", line 5301, in accept
  File "mypy/nodes.py", line 1021, in accept
  File "mypy/semanal.py", line 1121, in visit_class_def
  File "mypy/semanal.py", line 1201, in analyze_class
  File "mypy/semanal.py", line 1210, in analyze_class_body_common
  File "mypy/semanal.py", line 1256, in apply_class_plugin_hooks
  File "mypy/plugins/dataclasses.py", line 524, in dataclass_class_maker_callback
  File "mypy/plugins/dataclasses.py", line 120, in transform
  File "mypy/plugins/dataclasses.py", line 408, in collect_attributes
  File "mypy/plugins/dataclasses.py", line 106, in expand_typevar_from_subtype
  File "mypy/typeops.py", line 151, in map_type_from_supertype
  File "mypy/maptype.py", line 24, in map_instance_to_supertype
  File "mypy/maptype.py", line 37, in map_instance_to_supertypes
  File "mypy/maptype.py", line 82, in map_instance_to_direct_supertypes
  File "mypy/expandtype.py", line 18, in expand_type
  File "mypy/types.py", line 1127, in accept
  File "mypy/expandtype.py", line 90, in visit_instance
  File "mypy/expandtype.py", line 212, in expand_types
  File "mypy/types.py", line 2472, in accept
AssertionError: 
efax/_src/distributions/dirichlet.py:29: : note: use --pdb to drop into pdb
```

### PR DESCRIPTION
**Pull request by JukkaL**
The dataclass plugin could crash if it encountered a placeholder. Fix the issue by
running the plugin after the main semantic analysis pass, when all placeholders have 
been resolved.

Also add a new hook called `get_class_decorator_hook_2` that is used by the
dataclass plugin.

We may want to do a similar change to the attrs plugin, but let's change one thing
at a time.

Fix #12685.

### CODE REVIEW

File: mypy/plugin.py [hunk 1]
```diff
@@ -692,9 +692,32 @@ def get_class_decorator_hook(self, fullname: str
 
         The plugin can modify a TypeInfo _in place_ (for example add some generated
         methods to the symbol table). This hook is called after the class body was
-        semantically analyzed.
+        semantically analyzed, but *there may still be placeholders*.
```
Comment (ilevkivskyi) [inline mypy/plugin.py:695]: Maybe briefly mention that placeholders are caused by forward refs (plugin devs read this file as docs).

File: mypy/plugins/dataclasses.py [hunk 2]
```diff
@@ -390,6 +397,9 @@ def collect_attributes(self) -> Optional[List[DataclassAttribute]]:
         # we'll have unmodified attrs laying around.
         all_attrs = attrs.copy()
         for info in cls.info.mro[1:-1]:
+            if 'dataclass_tag' in info.metadata and 'dataclass' not in info.metadata:
+                # We haven't processed the base class yet. Need another pass.
```
Comment (ilevkivskyi) [inline mypy/plugins/dataclasses.py:401]: Maybe also add the meaning of returning `None` to the docstring?

File: mypy/semanal_main.py [hunk 3]
```diff
@@ -382,14 +387,62 @@ def check_type_arguments_in_targets(targets: List[FineGrainedDeferredNode], stat
...
+
+    Note that some hooks incorrectly run during the main semantic
+    analysis pass, for historical reasons.
+    """
+    num_passes = 0
+    incomplete = True
+    # If we encounter a base class that has not been processed, we'll run another
+    # pass. This should eventually reach a fixed point.
+    while incomplete:
+        assert num_passes < 5, "Internal error: too many class plugin hook passes"
```
Comment (ilevkivskyi) [inline mypy/semanal_main.py:405]: This looks a bit arbitrary. I would say we need a large number, so this will not happen in legitimate cases.

File: mypy/semanal_main.py [hunk 4]
```diff
@@ -382,14 +387,62 @@ def check_type_arguments_in_targets(targets: List[FineGrainedDeferredNode], stat
...
     for module in scc:
-        tree = graph[module].tree
+        state = graph[module]
+        tree = state.tree
         assert tree
         for _, node, _ in tree.local_definitions():
             if isinstance(node.node, TypeInfo):
-                saved = (module, node.node, None)  # module, class, function
-                with errors.scope.saved_scope(saved) if errors.scope else nullcontext():
+                with state.manager.semantic_analyzer.file_context(tree, state.options, node.node):
```
Comment (ilevkivskyi) [inline mypy/semanal_main.py:445]: Not sure why these changes are needed here.

**Review by ilevkivskyi — APPROVED** [review]

File: mypy/semanal_main.py (hunk 4 above)
Comment (JukkaL) [reply mypy/semanal_main.py:445]: Some errors were reported incorrectly in test cases (in the wrong file etc.) without these changes. I think that that this previously implicitly assumed that the context state was set up via some earlier pass, and the new pass invalidated some of the assumptions. Now we explicitly set the state to what we require.
